File: code-samples/ai/code-snippets/speaker-identification.java
package SpeakersIdentification;

import java.io.IOException;
import com.google.common.reflect.TypeToken;
import com.google.gson.Gson;

import com.ringcentral.*;
import com.ringcentral.definitions.*;

public class SpeakersIdentification {
    String NGROK_ADDRESS = "NGROK-TUNNEL-ADDRESS";
    String WEBHOOK_URL = NGROK_ADDRESS + "/webhook";
    String CONTENT_URI = "PUBLICLY-ACCESSIBLE-CONTENT-URI";

    static RestClient restClient;

    public static void main(String[] args) {
      var obj = new AnalyzeInteraction();
      try {
        // Instantiate the SDK
        restClient = new RestClient("PRODUCTION-APP-CLIENT-ID", "PRODUCTION-APP-CLIENT-SECRET", "https://platform.ringcentral.com");

        // Authenticate a user using a personal JWT token
        restClient.authorize("PRODUCTION-JWT");

        obj.speakers_identification();

      } catch (RestException e) {
        System.out.println(e.getMessage());
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
    /*
    * Identify speakers from a conversation
    */
    static private void speakers_identification() throws RestException, IOException {
        var enrolledSpeakerIds = read_enrolled_speakers();
        if (enrolledSpeakerIds.length > 0) {
            try
            {
                var bodyParams = new IdentifyInput()
                    .contentUri(CONTENT_URI)
                    .encoding("Wav")
                    .languageCode("en-US")
                    .source("RingCentral")
                    .audioType("CallCenter")
                    .enableVoiceActivityDetection(true)
                    .speakerIds(enrolledSpeakerIds);

                var queryParams = new CaiSpeakerIdentifyParameters().webhook(WEBHOOK_URL);
                var resp = restClient.ai().audio().v1().async().speakerIdentify().post(bodyParams, queryParams);
                System.out.println("Job ID: " + resp.jobId);
                System.out.println("Ready to receive response at: " + WEBHOOK_URL);
            }catch (RestException ex) {
            	System.out.println("Unable to call speaker identify API. " + ex.getMessage());
            }
        } else {
        	System.out.println("No enrolled speakers. Please enroll a few speaker ids and try again.");
        }
    }

    /*
    * Read the account enrolled speakers
    */
    static private String[] read_enrolled_speakers() throws RestException, IOException {
        ArrayList<String> speakerIdList = new ArrayList<>();
        try {
            var queryParams = new CaiEnrollmentsListParameters()
	            .partial(true)
                .perPage(100l)
                .page(1l);

            var resp = restClient.ai().audio().v1().enrollments().list(queryParams);
            for (var enrollment : resp.records) {
            	speakerIdList.add(enrollment.speakerId);
            }
        } catch (RestException ex) {
            System.out.println("Unable to find enrolled speakers. " + ex.getMessage());
        }
        String[] enrolledSpeakerIds = speakerIdList.toArray(new String[speakerIdList.size()]);
        return enrolledSpeakerIds;
    }
}

File: code-samples/ai/code-snippets/speaker-identification.js
const fs = require ('fs')
var platform = require('./../quick-start.js').platform;
NGROK_ADDRESS = process.env.NGROK_URL // "http://d973-73-170-11-87.ngrok-free.app"
WEBHOOK_URL = NGROK_ADDRESS + "/webhook";
CONTENT_URI = 'https://rclabs-addin-resources.s3.us-east-1.amazonaws.com/media/Phong-Swetha-01.mp3'
speakers_identification()


// Next line must be at the 10th line!
platform.on(platform.events.loginSuccess, () => {
    NGROK_ADDRESS = "NGROK-TUNNEL-ADDRESS"
    WEBHOOK_URL = NGROK_ADDRESS + "/webhook";
    CONTENT_URI = "PUBLICLY-ACCESSIBLE-CONTENT-URI"
    speakers_identification()
})

platform.on(platform.events.loginError, function(e){
    console.log("Unable to authenticate to platform. Check credentials.", e.message)
    process.exit(1)
});

/*
* Identify speakers from a conversation
*/
async function speakers_identification() {
    let enrolledSpeakerIds = await read_enrolled_speakers()
    if (enrolledSpeakerIds.length > 0){
      try {
        let bodyParams = {
            contentUri:   CONTENT_URI,
            encoding:     "Mpeg",
            languageCode: "en-US",
            source:       "RingCentral",
            audioType:    "CallCenter",
            speakerIds:   enrolledSpeakerIds
        }
        let endpoint = `/ai/audio/v1/async/speaker-identify?webhook=${WEBHOOK_URL}`
        let resp = await platform.post(endpoint, bodyParams);
        let jsonObj = await resp.json();
        if (resp.status == 202) {
          console.log("Job ID: " + jsonObj.jobId);
          console.log("Ready to receive response at: " + WEBHOOK_URL);
        }
      } catch (e) {
          console.log(`Unable to call speaker identify API. ${e.message}`);
      }
    }else{
      console.log("No enrolled speakers. Please enroll a few speaker ids and try again.")
    }
}

/*
* Read the account enrolled speakers
*/
async function read_enrolled_speakers() {
  var enrolledSpeakerIds = []
  try{
    let queryParams = {
        partial: false,
        perPage: 100,
        page: 1
    }
    let endpoint = "/ai/audio/v1/enrollments"
    var resp = await platform.get(endpoint, queryParams)
    var jsonObj = await resp.json()
    for (var enrollment of jsonObj.records){
      enrolledSpeakerIds.push(enrollment.speakerId)
    }
  }catch (e){
    console.log("Unable to find enrolled speakers.", e.message)
  }
  return enrolledSpeakerIds
}
